lib/rethinkdb/connection/multiplexer.ex

```diff
@@ -0,0 +1,214 @@
+defmodule RethinkDB.Connection.Multiplexer do
+  @moduledoc """
+  This module provides functions to multiplex messages over a single `:gen_tcp`
+  connection.
+
+  Each request message sent to the server has a unique 8-byte token assigned. The
+  server will send responses using this token as an identifier so the reply can
+  be matched to its request.
+
+  Each message is encoded in following format:
+
+  * The 8-byte unique message token.
+  * The length of the message, as a 4 byte little-endian integer.
+  * The message content.
+
+  The advantage of using this approach is that it allows the server to process
+  multiple requests in parallel on the same connection. For the client, this
+  means that replies can arrive in any order.
+  """
+  use GenServer
+
+  import DBConnection.ConnectionError, only: [exception: 1]
+
+  @doc false
+  defstruct conn: {nil, nil}, token: 1, ets: nil, ref: nil
+
+  @doc """
+  Starts a `Connection` process linked to the current process.
+  """
+  def start_link(options \\ []) do
+    GenServer.start_link(__MODULE__, options, [])
+  end
+
+  @doc """
+  Sends the given `message` to the server.
+
+  Options:
+
+  * `:timeout` - maximum time to wait for a reply (default: `5_000`).
+  * `:noreply` - do not wait for the server acknowledgement of the message.
+  """
+  def send_recv(pid, message, options \\ []) do
+    noreply = Keyword.get(options, :noreply, false)
+    timeout = Keyword.get(options, :timeout, 5_000)
+    if noreply do
+      GenServer.cast(pid, {:send, message})
+    else
+      token = Keyword.get(options, :token)
+      {:ok, GenServer.call(pid, {:send, token, message}, timeout)}
+    end
+  end
+
+  #
+  # GenServer API
+  #
+
+  def init(options) do
+    host = Keyword.get(options, :hostname, "localhost")
+    port = Keyword.get(options, :port, 28015)
+    user = Keyword.get(options, :user, "admin")
+    pass = Keyword.get(options, :password, "")
+
+    {transport, sock_opts} =
+      if ssl_opts = Keyword.get(options, :ssl) do
+        {:ssl, Enum.map(Keyword.fetch!(ssl_opts, :ca_certs),  &({:cacertfile, &1})) ++ [verify: :verify_peer]}
+      else
+        {:gen_tcp, []}
+      end
+
+    # Connects to the server and perform a handshake
+    sock_opts = [packet: :raw, mode: :binary, active: :false] ++ sock_opts
+    with {:ok, sock} <- transport.connect(String.to_char_list(host), port, sock_opts),
+         {:ok, conn} <- init_connection(transport, sock),
+         {:ok, _nil} <- handshake(conn, user, pass),
+         {_pid, ref} <- spawn_monitor(__MODULE__, :recv_loop, [self, conn]),
+    do:  {:ok, %__MODULE__{conn: conn, ets: :ets.new(__MODULE__, [:set, :private]), ref: ref}}
+  end
+
+  def handle_call({:send, nil, msg}, from, %__MODULE__{conn: conn, token: token, ets: table} = state) do
+    :ok = send_payload(conn, token, msg)
+    :ets.insert(table, {token, from})
```

Comment:
Why use `:ets`? Why not just use the process state? The advantage of `:ets` is you can have readers from outside of this process (which we do not have).